### DISCUSSION TITLE
FFMPEG CPU usage high on Khadas ARM (with Video acceleratoin)

### DISCUSSION THREAD

**Discussion by RichardPar** [blakeblackshear/frigate#6053] [General]
Hi there, I have just enabled video acceleration on my platform and I am noticing that CPU usage is high on FFMPEG

Running the command directly, I am getting 3.9% - Frigate command uses 30%

Is there a tweak to make it not burn up the CPU?  

```
khadas@Khadas:~$ ps -p 22916 -o %cpu,%mem,cmd
%CPU %MEM CMD
 3.9  0.5 ffmpeg -c:v h264_v4l2m2m -i rtsp://localhost:8554/garden_low_cam -f null /dev/null
```
The output of the Frigate process shows 30%.  I am aware its doing more - 

```
khadas@Khadas:~$ ps -p 10743 -o %cpu,%mem,cmd
%CPU %MEM CMD
30.7  0.7 ffmpeg -hide_banner -loglevel warning -c:v h264_v4l2m2m -user_agent FFmpeg Frigate/0.12.0-da3e197 -avoid_negative_ts make_zero -fflags +genpts+discardc
khadas@Khadas:~$
```
(Full command)  
```
ffmpeg -hide_banner -loglevel warning -c:v h264_v4l2m2m -user_agent FFmpeg Frigate/0.12.0-da3e197 -avoid_nega                                       tive_ts make_zero -fflags +genpts+discardcorrupt -rtsp_transport tcp -stimeout 5000000 -use_wallclock_as_timestamps 1 -i rtsp://127.0.0.1:8554/garden_low_cam -r 10 -s 640x352 -threads 2 -f rawvideo -pix_fmt yuv420p pipe: 
```

**Comment by RichardPar** [comment]
On digging...

The Pix Format is the bit killing the CPU on the output from a hardware decoder '-pix_fmt yuv420p'

**Reply by NickM-27** [reply]
yes, pixel format change is not done on GPU even when hwaccel is specified. This is expected for now as frigate works with yuv420p

**Reply by RichardPar** [reply]
> yes, pixel format change is not done on GPU even when hwaccel is specified. This is expected for now as frigate works with yuv420p

I am a bit puzzled... the output from the hardware decoder is already yuv420p

```
Input #0, rtsp, from 'rtsp://localhost:8554/garden_low_cam':
  Metadata:
    title           : go2rtc/1.2.0
  Duration: N/A, start: 0.100000, bitrate: N/A
    Stream #0:0: Video: h264 (Main), yuv420p(tv, bt470bg/bt470bg/smpte170m, progressive), 640x352, 10 tbr, 90k tbn, 180k tbc
Stream mapping:
  Stream #0:0 -> #0:0 (h264 (native) -> wrapped_avframe (native))
Press [q] to stop, [?] for help
Output #0, null, to '/dev/null':
  Metadata:
    title           : go2rtc/1.2.0
    encoder         : Lavf58.45.100
    Stream #0:0: Video: wrapped_avframe, yuv420p, 640x352, q=2-31, 200 kb/s, 10 fps, 10 tbn, 10 tbc
    Metadata:
      encoder         : Lavc58.91.100 wrapped_avframe
frame=   65 fps= 14 q=-0.0 Lsize=N/A time=00:00:07.80 bitrate=N/A speed=1.74x
video:34kB audio:0kB subtitle:0kB other streams:0kB global headers:0kB muxing overhead: unknown
```

**Reply by NickM-27** [reply]
I am not sure on your particular hardware decoder but most GPU decoders only work in nv12 format so it needs conversions, and that is why

**Reply by RichardPar** [reply]
> I am not sure on your particular hardware decoder but most GPU decoders only work in nv12 format so it needs conversions, >and that is why

This is a hardware v4l decoder (like raspberry, but not!) 

 I added 'rawvideo' and the output is listed as I420 (which in my head is the same as yuv420p) 
``` 
Output #0, rawvideo, to '/dev/null':
  Metadata:
    title           : go2rtc/1.2.0
    encoder         : Lavf58.45.100
    Stream #0:0: Video: rawvideo (I420 / 0x30323449), yuv420p, 640x352, q=2-31, 27033 kb/s, 10 fps, 10 tbn, 10 tbc
    Metadata:
      encoder         : Lavc58.91.100 rawvideo
frame=   87 fps= 28 q=-0.0 Lsize=   28710kB time=00:00:09.00 bitrate=26132.5kbits/s speed=2.92x
video:28710kB audio:0kB subtitle:0kB other streams:0kB global headers:0kB muxing overhead: 0.000000%
Exiting normally, received signal 2.
```

**Reply by RichardPar** [reply]
> I am not sure on your particular hardware decoder but most GPU decoders only work in nv12 format so it needs conversions, and that is why

You are right.. it is NV12 
```
Output #0, rawvideo, to '/dev/null':
  Metadata:
    title           : go2rtc/1.2.0
    encoder         : Lavf58.45.100
    Stream #0:0: Video: rawvideo (NV21 / 0x3132564E), nv21, 640x384 [SAR 1:1 DAR 5:3], q=2-31, 73728 kb/s, 25 fps, 25 tbn, 25 tbc
    Metadata:
      encoder         : Lavc58.91.100 rawvideo
^Came=  173 fps= 50 q=-0.0 size=   62208kB time=00:00:08.24 bitrate=61845.6kbits/s dup=103 drop=0 speed=2.36x
```

**Comment by RichardPar** [comment]
Fixed..

On the Khadas VIM4 boards, you add another parameter to the hwaccel and removed the colourspace conversion on the output
```
-c:v:1   <--------- :1 is added.  Using :0 is NV12 format
```

```
ffmpeg:
   global_args: -hide_banner -loglevel debug
   input_args: preset-rtsp-generic
   hwaccel_args:
     - -c:v:1
     - h264_v4l2m2m   
   output_args:
     detect: -threads 2 -f rawvideo
```

**Reply by blakeblackshear** [reply]
If you use the default yuv420 in the output args, does it hurt anything? Now that the formats match, I don't think it will do any extra work with the default args.

**Reply by RichardPar** [reply]
may not... but if its not needed, I dont see the point of it being there :D

**Reply by timotto** [reply]
The exact same configuration helped to bing CPU usage down from 90% to 30% for a 1080p stream on my Raspberry Pi 4.

Without the `-c:v:1` parameter the hardware decoder outputs `NV12`. Now it outputs `I420`:
```
Input:
    Stream #0:0: Video: h264 (High), yuvj420p(pc, bt709, progressive), 1920x1080, 20 fps, 20 tbr, 90k tbn, 40 tbc
Output:
    Stream #0:0: Video: rawvideo (I420 / 0x30323449), yuvj420p, 1920x1080, q=2-31, 124416 kb/s, 5 fps, 5 tbn, 5 tbc
```

if I add `-pix_fmt yuv420p` to the output args the CPU usage jumps up to 90% again, probably because `yuvj420p` isn't **exactly** `yuv420p`. I can't see any difference in image quality if I use `-pix_fmt yuv420p` or not. The extra color range of `yuvj420p` does not cause any overflow/flipped colors even in pitch black or bright sunlight. Detection works great.

**Reply by RichardPar** [reply]
> The exact same configuration helped to bing CPU usage down from 90% to 30% for a 1080p stream on my Raspberry Pi 4.
> 
> Without the `-c:v:1` parameter the hardware decoder outputs `NV12`. Now it outputs `I420`:
> 
> ```
> Input:
>     Stream #0:0: Video: h264 (High), yuvj420p(pc, bt709, progressive), 1920x1080, 20 fps, 20 tbr, 90k tbn, 40 tbc
> Output:
>     Stream #0:0: Video: rawvideo (I420 / 0x30323449), yuvj420p, 1920x1080, q=2-31, 124416 kb/s, 5 fps, 5 tbn, 5 tbc
> ```
> 
> if I add `-pix_fmt yuv420p` to the output args the CPU usage jumps up to 90% again, probably because `yuvj420p` isn't **exactly** `yuv420p`. I can't see any difference in image quality if I use `-pix_fmt yuv420p` or not. The extra color range of `yuvj420p` does not cause any overflow/flipped colors even in pitch black or bright sunlight. Detection works great.

try other values .. ```-c:v:0``` ```-c:v:1``` ```-c:v:2``` etc etc
